File: frontend/src/visualizers/wave.ts
import type { DrawFn } from './types'
import { makeGradient, applyGlow, getWaveSamples, hexToRgba } from './types'

export const drawWave: DrawFn = ({ ctx, W, H, t, audio, viz }) => {
  const { colorMode, primaryColor, secondaryColor, opacity,
          glow, sensitivity, posX, posY, scale, mirror } = viz

  const samples  = getWaveSamples(audio, 512, t)
  const cx       = W * (posX / 100)
  const cy       = H * (posY / 100)
  const waveW    = W * 0.9 * scale
  const waveH    = H * 0.30 * scale
  const startX   = cx - waveW / 2

  ctx.save()
  applyGlow(ctx, primaryColor, glow)

  const strokeColor = colorMode === 'solid'
    ? hexToRgba(primaryColor, opacity)
    : makeGradient(ctx, startX, 0, startX + waveW, 0, primaryColor, secondaryColor, opacity)

  function drawPath(flip: boolean, alphaScale = 1.0) {
    ctx.beginPath()
    for (let i = 0; i < samples.length; i++) {
      const x   = startX + (i / (samples.length - 1)) * waveW
      const raw = samples[i] * sensitivity
      const y   = cy + (flip ? -raw : raw) * waveH * 0.45
      i === 0 ? ctx.moveTo(x, y) : ctx.lineTo(x, y)
    }
    ctx.strokeStyle = strokeColor
    ctx.lineWidth   = 2.5
    ctx.globalAlpha = alphaScale
    ctx.stroke()
  }

  drawPath(false)
  if (mirror) drawPath(true, 0.45)

  // Filled area below the wave
  ctx.beginPath()
  ctx.moveTo(startX, cy)
  for (let i = 0; i < samples.length; i++) {
    const x   = startX + (i / (samples.length - 1)) * waveW
    const raw = samples[i] * sensitivity
    const y   = cy + raw * waveH * 0.45
    ctx.lineTo(x, y)
  }
  ctx.lineTo(startX + waveW, cy)
  ctx.closePath()
  const fillGrad = ctx.createLinearGradient(0, cy - waveH, 0, cy + waveH)
  fillGrad.addColorStop(0, hexToRgba(primaryColor, opacity * 0.3))
  fillGrad.addColorStop(1, hexToRgba(primaryColor, 0))
  ctx.fillStyle = fillGrad
  ctx.globalAlpha = 1
  ctx.shadowBlur = 0
  ctx.fill()

  ctx.restore()
}

File: backend/app/services/visualizers/wave.py
"""Port of frontend/src/visualizers/wave.ts"""
from __future__ import annotations

from typing import Optional

import numpy as np
from PIL import Image, ImageDraw

from app.services.audio_analyzer import AudioFrame
from app.services.visualizers.types import (
    apply_glow, get_wave_samples, hex_to_rgba, lerp_rgba, _grad_h,
)


def draw_wave(
    W: int, H: int, t: float,
    audio: Optional[AudioFrame],
    s,
) -> Image.Image:
    layer = Image.new("RGBA", (W, H), (0, 0, 0, 0))
    draw  = ImageDraw.Draw(layer)

    samples = get_wave_samples(audio, 512, t)
    n       = len(samples)
    cx      = W * (s.pos_x / 100)
    cy      = H * (s.pos_y / 100)
    wave_w  = W * 0.9 * s.scale
    wave_h  = H * 0.30 * s.scale
    start_x = cx - wave_w / 2

    pts = []
    for i in range(n):
        x   = start_x + (i / (n - 1)) * wave_w
        raw = samples[i] * s.sensitivity
        y   = cy + raw * wave_h * 0.45
        pts.append((x, y))

    # Stroke — gradient drawn segment by segment
    a = round(s.opacity * 255)
    if s.color_mode == "gradient":
        for i in range(len(pts) - 1):
            frac = i / (n - 1)
            color = lerp_rgba(s.primary_color, s.secondary_color, frac, s.opacity)
            draw.line([pts[i], pts[i + 1]], fill=color, width=3)
    else:
        color = hex_to_rgba(s.primary_color, s.opacity)
        draw.line(pts, fill=color, width=3)

    if s.mirror:
        mir_pts = [(x, cy - (y - cy)) for x, y in pts]
        mir_col = hex_to_rgba(s.primary_color, s.opacity * 0.45)
        draw.line(mir_pts, fill=mir_col, width=2)

    # Filled area below the wave
    poly = [(start_x, cy)] + pts + [(start_x + wave_w, cy)]
    fill_col = hex_to_rgba(s.primary_color, s.opacity * 0.15)
    draw.polygon(poly, fill=fill_col)

    if s.glow > 0:
        layer = apply_glow(layer, s.glow)

    return layer
